core/src/output_writer/hdf5/hdf5.h

```diff
@@ -0,0 +1,167 @@
+  //! combinedMeshesOut contains the 1D mesh names that were written to the hdf5
+  //! file.
+  template <typename FieldVariablesForOutputWriterType>
+  void
+  writePolyDataFile(hid_t fileID,
+                    const FieldVariablesForOutputWriterType &fieldVariables,
+                    std::set<std::string> &combinedMeshesOut);
+
+  //! write all data of all 3D or 2D field ! variables (depending on
+  //! output3DMeshes) into a group. This is uses MPI IO. It can be enabled with
+  //! the "combineFiles" option. on return, combinedMeshesOut contains the 3D or
+  //! 2D mesh names that were written to the vtu file.
+  template <typename FieldVariablesForOutputWriterType>
+  void writeCombinedUnstructuredGridFile(
+      hid_t fileID, const FieldVariablesForOutputWriterType &fieldVariables,
+      std::set<std::string> &combinedMeshesOut, bool output3DMeshes);
+
+protected:
+  /** one Piece is the file output. It is created from one or multiple opendihu
+   * meshes
+   */
+  struct Piece {
+    std::set<std::string>
+        meshNamesCombinedMeshes; //< the meshNames of the combined meshes, or
+                                 // only one meshName if it is not a merged mesh
+    std::vector<std::string>
+        meshNamesCombinedMeshesVector; //< the same as meshNamesCombinedMeshes,
+                                       // but as vector that preserves the
+                                       // order, this is important for the
+                                       // output file
+    PolyDataPropertiesForMesh properties; //< the properties of the merged mesh
+
+    std::string firstScalarName; //< name of the first scalar field variable of
+                                 // the mesh
+    std::string
+        firstVectorName; //< name of the first vector field variable with 3
+                         // components of the mesh
+
+    //! constructor, initialize nPoints and nCells to 0
+    Piece();
+
+    //! assign the correct values to firstScalarName and firstVectorName, only
+    //! if properties has been set
+    void setVTKValues();
+  };
+
+  //! write the values vector combined to the file, correctly encoded,
+  //! identifier is an id to access cached values
+  template <typename T>
+  void writeCombinedValuesVector(hid_t fileID, const std::vector<T> &values,
+                                 const char *dsname,
+                                 bool writeFloatsAsInt = false);
+
+  //! write a vector containing nValues "12" (if output3DMeshes) or "9" (if
+  //! !output3DMeshes) values for the types for an unstructured grid
+  void writeCombinedTypesVector(hid_t fileHandle, int nValues,
+                                bool output3DMeshes, const char *dsname);
+
+private:
+  //! open a HDF5 file with a given filename
+  hid_t openHDF5File(const char *filename, bool mpiio);
+
+  //! helper method that writes the unstructured grid file
+  template <typename FieldVariablesForOutputWriterType>
```

Comment:
problem here is that i would not be able to call it if it would be part of a cpp file because i need to call it from within `innerWrite` which is a template function (has to be a template function) and that function needs to be part of the header file (or more specific the tpp file here)

Reply:
Ahh I forgot we use *.tpp files. In that case lets keep it this way.